src/index.ts

```diff
@@ -248,6 +256,42 @@ export const supportedV2RouteHandler = (req, res, next, server) => {
   return server(req, res, next)
 }
 
+// Middleware that must be mounted on our GraphQL route _before_
+// our regular GraphQL handler. This is so we can register a new `res.end`
+// method that will intercept the response and check if any email provider
+// rate limting occurred.
+//
+// If so, we return a 429 status code and a JSON error message.
+app.use("/v2", function (req, res, next) {
+  const userAgent = req.headers["user-agent"] || ""
+  if (!userAgent.includes("Braze")) {
+    return next()
+  }
+
+  const originalEnd = res.end
+
+  res.end = function (...args) {
+    res.end = originalEnd
+    const [chunk, encoding, cb] = args
+
+    const contentType = res.get("content-type") || "text/html"
+
+    if (contentType.indexOf("application/json") === 0) {
+      const response = JSON.parse(chunk)
```

Comment:
It's unfortunate we have to re-parse this.

Uninformed question: I've seen `context` used for passing request data around. Is it not possible to use that to make the 429 status available, as opposed to rendering it into the extensions response?

Reply:
I suppose one _could_ use `context` for that, but that's a GraphQL-server-concept (as in, within the scope of the special `express-graphql` sub-app mounted at our /v2 endpoint) - and so _this_ middleware (which is a totally separate Express middleware mounted before that sub-app) _doesn't_ have access to that `context`.

In general trying to implement some version of this was tricky for exactly this reason!

- Can't do it natively with `express-graphql` as there's no `parseResponse` or any other 'hook' available -> leading to possible patch.
- Can't do it natively with Express middleware as that won't have access to the completed response, unless there's some tricksy `res.on("end", ...` overriding - and then this approach implies that we _do_ need to store the 'this is an email provider request that resulted in rate limiting' fact in the response somewhere.